geopy/geocoders/here.py

```diff
@@ -0,0 +1,281 @@
+"""
+:class:`.Here` geocoder.
+"""
+
+from geopy.compat import urlencode
+from geopy.exc import (
+    GeocoderAuthenticationFailure,
+    GeocoderInsufficientPrivileges,
+    GeocoderQuotaExceeded,
+    GeocoderServiceError,
+    GeocoderUnavailable,
+)
+from geopy.geocoders.base import DEFAULT_SENTINEL, Geocoder
+from geopy.location import Location
+from geopy.util import join_filter, logger
+
+__all__ = ("Here", )
+
+
+class Here(Geocoder):
+    """Geocoder using the HERE Geocoder API.
+
+    Documentation at:
+        https://developer.here.com/documentation/geocoder/
+    """
+
+    structured_query_params = {
+        'city',
+        'county',
+        'district',
+        'country',
+        'state',
+        'street',
+        'housenumber',
+        'postalcode',
+    }
+
+    def __init__(
+            self,
+            app_id,
+            app_code,
+            format_string=None,
+            scheme=None,
+            timeout=DEFAULT_SENTINEL,
+            proxies=DEFAULT_SENTINEL,
+            user_agent=None,
+            ssl_context=DEFAULT_SENTINEL,
+    ):
+        """
+
+        :param str app_id: Should be a valid HERE Maps APP ID
+            See https://developer.here.com/authenticationpage.
+
+        :param str app_code: Should be a valid HERE Maps APP CODE
+            See https://developer.here.com/authenticationpage.
+
+        :param str format_string:
+            See :attr:`geopy.geocoders.options.default_format_string`.
+
+        :param str scheme:
+            See :attr:`geopy.geocoders.options.default_scheme`.
+
+        :param int timeout:
+            See :attr:`geopy.geocoders.options.default_timeout`.
+
+        :param dict proxies:
+            See :attr:`geopy.geocoders.options.default_proxies`.
+
+        :param str user_agent:
+            See :attr:`geopy.geocoders.options.default_user_agent`.
+
+        :type ssl_context: :class:`ssl.SSLContext`
+        :param ssl_context:
+            See :attr:`geopy.geocoders.options.default_ssl_context`.
+        """
+        super(Here, self).__init__(
+            format_string=format_string,
+            scheme=scheme,
+            timeout=timeout,
+            proxies=proxies,
+            user_agent=user_agent,
+            ssl_context=ssl_context,
+        )
+        self.app_id = app_id
+        self.app_code = app_code
+        self.api = "%s://geocoder.api.here.com/6.2/geocode.json" % self.scheme
+        self.reverse_api = "%s://reverse.geocoder.api.here.com/6.2/reversegeocode.json" \
+            % self.scheme
+        # FIXME(?): add multi-reverse geocode
+        # FIXME(?): add landmark geocode
+
+    def geocode(
+            self,
+            query,
+            exactly_one=True,
+            language=None,
+            additional_data=False,
+            other=None,
```

Comment:
Could you please remove this kwarg? As stated in [CONTRIBUTING.md][1], allowing to pass arbitrary parameters is discouraged, as it's prone to break users' code on API version upgrades.

[1]: https://github.com/geopy/geopy/blob/602150a3c9c4ec63206325db4e50a3c9b40b43ac/CONTRIBUTING.md#improving-a-geocoder

Reply:
I totally understand the motivation. From all the existing params I've taken only `language` as the one available for most/all geocoders (I guess) and `additional_data` as this allows for getting shapes, too, which I think is rather unusual. The problem is: there are *many* more params, and only `additional_data` can have [17 different keys](https://developer.here.com/documentation/geocoder/topics/resource-params-additional.html). For the geocode API there are [28 more](https://developer.here.com/documentation/geocoder/topics/resource-geocode.html), plus [21 for the reverse geocode API](https://developer.here.com/documentation/geocoder/topics/resource-reverse-geocode.html)! And I haven't even added the multi-reverse geocode and landmark geocode APIs... (see FIXMEs).

So I felt like having an `other` parameter is the most efficient way of getting something out quickly, being well aware that this can not be documented nicely inside geopy, but only on the external REST API documentation. In this case I was told this is unlikely to change since there will be a new generation of these HERE APIs under a different label, anyway, but I'll double-check, again. So, if possible I'd really avoid putting in such a huge amount of work, but instead try to add a few more interesting test cases. Would that be an option? How is that handled for other rich APIs?

Reply:
A lot of geocoding services provide a rich set of parameters, options and API methods. But more than 28 parameters indeed look rather unusual though.

Generally implementing full support for all parameters and methods is not a goal in geopy. Implementing basic forward/reverse geocoding is usually enough for a lot of use-cases. For missing features [the doc suggests][semver] to open PRs for adding those (yup, the wording might have been more explicit with regards to missing API parameters).

While it's possible to explicitly define all of these peculiar parameters, a really small subset of users would really need them. And even then they would probably be better off using a specialized SDK package for that geocoder, which goal is exactly to match the API as close as possible. But, having all these rarely needed parameters implemented in geopy, would force us to rename the geocoder class once they change their API version (like it was done with `Google` and `GoogleV3` a while ago), if they drop or change some of the parameters; in order to not silently break the code of people who did use these parameters. But the large group of people, who didn't, would have to rename the class to upgrade, while other changes from them are not needed.

So the general approach is trying to keep the geocoder's interface sufficient enough for typical use-cases. Easier API version upgrades is more of a consequence of that rather than a goal.

The exact set of parameters is up to you. For me what's already there looks fine.

[semver]: https://geopy.readthedocs.io/en/1.14.0/#semver

Reply:
Well, I can look up some more that might be useful more generally. The point is that allowing only a specific subset makes it *impossible* to use the rest, while using a more generic approach only makes it *harder* to document. I would favour the first giving people more flexibility, but it's your package. :)

Reply:
Not impossible exactly. I can come up with at least one dirty workaround involving `threading.local()` and overriding of `_call_geocoder`. This is *very* ugly, but possible. :)

Please don't get me wrong: I would love to give that flexibility. But not at a cost of painful upgrades for the most of the users. It's definitely not just a matter of documentation. It's a matter of losing confidence over the scope of provided geocoder capabilities, which is crucial for preserving backwards compatibility when upgrading underlying APIs.

I've had this issue in mind for a long time, but now as you brought it, I've just created #305 to track this. I hope eventually we will come up with a decent solution to this.